### DISCUSSION TITLE
[3.0.0] Json Schemas for bindings and definitions current state

### DISCUSSION THREAD

**Discussion by Pakisan** [asyncapi/asyncapi-node#546] [General]
Let's discuss current state of Json Schemas and theirs validation trough tests

Ref: https://github.com/asyncapi/spec-json-schemas/pull/540

**Comment by Pakisan** [comment]
According to definition, [Reference](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#referenceObject) can't be extended

Motivation:
> This object cannot be extended with additional properties and any properties added SHALL be ignored.

Current Reference tests skip extensions part, because schema doesn't have `"additionalProperties": false`

So, let's add this restriction, or not?

@fmvilas @derberg @dalelane @smoya @char0n @GreenRover @jonaslagoni

**Reply by jonaslagoni** [reply]
> Current Reference tests skip extensions part, because schema doesn't have "additionalProperties": false

You should not be able to have additional properties alongside `$ref` no 🙂 So that should be fixed IMO 👍

**Comment by Pakisan** [comment]
According to definition, [Server](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#serverObject) can't be extended

Motivation:
Extensions were not mentioned

So, let's add this restriction, or not?

@fmvilas @derberg @dalelane @smoya @char0n @GreenRover @jonaslagoni

**Reply by jonaslagoni** [reply]
Good catch and question 👍

Up to the spec owners :v: But for now we could definitely enforce.

**Reply by smoya** [reply]
I believe this is an editorial error and everyone expects extensions to work at Server Object. Extensions were allowed in v2 as you can see in https://v2.asyncapi.com/docs/reference/specification/v2.6.0#serverObject

**Reply by Pakisan** [reply]
@smoya yeah, you are right

It's better to double check it

**Reply by smoya** [reply]
Feel free to open an issue in spec site mentioning this 🙏

**Reply by Pakisan** [reply]
> Feel free to open an issue in spec site mentioning this 🙏 

Yeah, here I collect founded issues with schema to discuss and open issues in spec repo, too

I think I'll finish review in a week, not much free time, on this week

**Comment by Pakisan** [comment]
BindingsObject:
- doesn't know about deprecated bindings like mqtt5
- doesn't recognize old versions of bindings

I propose to refactor (channel | server | operation | message)BindingsObject

**Comment by Pakisan** [comment]
Bindings:
- channel: SQS, Solace, WebSockets
- operation: Solace

Looks like they MUST respect this - This object MAY be extended with Specification Extensions.

**Comment by Pakisan** [comment]
AMQP channel binding

Looks like `is` MUST be required because it defines parsing strategy

**Comment by Pakisan** [comment]
IBM MQ operation binding

Looks like `type` MUST be required because it defines parsing strategy

**Comment by Pakisan** [comment]
IBM MQ channel binding

Looks like `destinationType` MUST be required because it defines parsing strategy

**Comment by Pakisan** [comment]
Amazon SNS channel binding

broken example for `filterPolicy` and broken `type` type

```json
[
  {
    "instancePath": "/consumers/0/filterPolicy/event/0",
    "schemaPath": "#/properties/filterPolicy/additionalProperties/oneOf/0/items/type",
    "keyword": "type",
    "params": { "type": "string" },
    "message": "must be string"
  },
  {
    "instancePath": "/consumers/0/filterPolicy/event",
    "schemaPath": "#/properties/filterPolicy/additionalProperties/oneOf/1/type",
    "keyword": "type",
    "params": { "type": "string" },
    "message": "must be string"
  },
  {
    "instancePath": "/consumers/0/filterPolicy/event",
    "schemaPath": "#/properties/filterPolicy/additionalProperties/oneOf/2/type",
    "keyword": "type",
    "params": { "type": "object" },
    "message": "must be object"
  },
  {
    "instancePath": "/consumers/0/filterPolicy/event",
    "schemaPath": "#/properties/filterPolicy/additionalProperties/oneOf",
    "keyword": "oneOf",
    "params": { "passingSchemas": null },
    "message": "must match exactly one schema in oneOf"
  }
]
```

**Comment by Pakisan** [comment]
Message Example not extendable now

But: This object MAY be extended with Specification Extensions.

**Comment by Pakisan** [comment]
Operation not extendable now

But: This object MAY be extended with Specification Extensions.

**Comment by Pakisan** [comment]
ReferenceObject:

error:
```json
{
    "instancePath": "/messages/0/$ref",
    "schemaPath": "http://asyncapi.com/definitions/3.0.0/ReferenceObject.json/format",
    "keyword": "format",
    "params": { "format": "uri-reference" },
    "message": "must match format \"uri-reference\""
  }
```

Value can't be:

```json
{
    "$ref": "#/components/{type}"
}
```

Only:

```json
{
    "$ref": "http://example.com"
}
```

Allow not inly uri

**Comment by Pakisan** [comment]
Looks like `Info Extensions` cannot be extended
